libraries/AP_BattMonitor/AP_BattMonitor_Scripting.h

```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+#include "AP_BattMonitor_Backend.h"
+
+#if HAL_BATTMON_SCRIPTING_ENABLED
+
+#define AP_BATTMONITOR_SCRIPTING_TIMEOUT_MICROS (5000000) // backend becomes unhealthy script does not call handler for 5 seconds
+
+class AP_BattMonitor_Scripting : public AP_BattMonitor_Backend
+{
+public:
+    // Constructor
+    AP_BattMonitor_Scripting(AP_BattMonitor &mon,
+                          AP_BattMonitor::BattMonitor_State &mon_state,
+                          AP_BattMonitor_Params &params);
+
+    bool has_time_remaining() const override { return _has_time_remaining; }
+    bool has_consumed_energy() const override { return _has_current; } // If we have current, we can report consumed energy. The internal flag _has_consumed_energy indicates whether the script overrides that calculation
+    bool has_current() const override { return _has_current; }
+    bool has_cell_voltages() const override { return _has_cell_voltages; }
+    bool has_temperature() const override { return _has_temperature; }
+    bool capacity_remaining_pct(uint8_t &percentage) const override;
+    bool get_cycle_count(uint16_t &cycles) const override;
+    AP_BattMonitor::Failsafe update_failsafes(void) override;
+    uint32_t get_mavlink_fault_bitmask() const override { return mavlink_faults;}
+    void update_resistance_estimate() override;
+
+    void init(void) override {};
+    void read() override;
+
+    bool handle_scripting(const BattMonitorScript_State &battmon_state) override;
+
+protected:
+    // These flags are updated when one of these states is changed from its
+    // invalid initial state magic value
+    bool _has_time_remaining = false;
+    bool _has_consumed_energy = false;
+    bool _has_current = false;
+    bool _has_cell_voltages = false;
+    bool _has_temperature = false;
+    bool _has_capacity_remaining_pct = false;
+    bool _has_cycle_count = false;
+
+    uint32_t mavlink_faults = 0;
```

Comment:
Don't need to init to 0.

Reply:
I'm a little confused by this. I thought nothing initializes to zero automatically in C++. Or are you saying I don't need to initialize because the script should set this on its own in the first loop, thus it's not needed to be initialized here? The only variable the script is required to set is voltage, so I want to make sure all other values initialize.

Reply:
Cool, I'll make the changes then. Could you point me in the general direction of how that happens in ArduPilot? Eager to learn what's going on under the hood there.

Reply:
Correct, One place is here, https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Common/c%2B%2B.cpp. We use calloc rather than malloc. We also do similar things in the various hal's to override malloc.